### ISSUE TITLE
emacsNativeComp: nearly half native-compiled eln files for builtin libs are missing

### ISSUE THREAD

**Issue by jian-lin** [NixOS/nixpkgs#170426]
### Describe the bug
The issue is what the title says.
By half, I mean

``` bash
$ ls /nix/store/waklcv94vzffa36qpfghs8c1yd0080wk-emacs-28.1/lib/emacs/28.1/native-lisp/28.1-8618a055 | wc -l
1349
```

and for my previously used `emacsGcc` from emacs-overlay

``` bash
$ ls /nix/store/hndnv6w8bzmw7vi5b1rcbkwf68la8lmy-emacs-gcc-28.0.92/lib/emacs/28.0.92/native-lisp/28.0.92-6a100051 | wc -l
2727
```

### Steps To Reproduce

1. add `emacsNativeComp` to `environment.systemPackages`
2. rebuild your system
3. clean the eln cache in your home, e.g. `mv ~/.emacs.d/eln-cache ~/.emacs.d/eln-cache.bak`
4. use `emacs -q` to open emacs
5. look for the manual for `cl-pushnew` by `M-x describe-function` and inputing `cl-pushnew`
6. check the eln cache in your home: `ls ~/.emacs.d/eln-cache/28.1*/`, you will find `cl-lib-*.eln` together other eln cache. I think this is a bug since `cl-lib` is builtin for emacs and it's eln cache should be distributed together with `emacsNativeComp`.

### Expected behavior
Uses of `emacsNativeComp` will not need to native-compile builtin libraries themselves.

### Additional context
`cl-lib-*.eln` does distributed together with `emacsGcc` from emacs-overlay I used before.

It's `/nix/store/hndnv6w8bzmw7vi5b1rcbkwf68la8lmy-emacs-gcc-28.0.92/lib/emacs/28.0.92/native-lisp/28.0.92-6a100051/cl-lib-8b938900-82325715.eln`.

### Notify maintainers
@lovek323 @jwiegley @adisbladis 
might be related to the bump of 28 #168076 , so @azahi 

### Metadata
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.34, NixOS, 22.05 (Quokka), 22.05.20220426.e120fd5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.0pre20220411_f7276bc`
 - nixpkgs: `/nix/store/xjycb37n2mxmhw03fy75xrz7lrlc0wrc-source`
```

**Comment by collares** [comment]
The difference between the two versions is that one is built from the tarball, while the other is built from Git. The tarball includes byte-compiled (`.elc`) files but not native-compiled (`.eln`) files. The build notices the `.elc` files and does not rebuild them, but native-compilation is a side-effect of byte-compilation and so the `.eln` files aren't built either.

(Almost all of the `.eln` files you see for the new version are trampolines, there are only 3 non-trampoline files.)

**Comment by collares** [comment]
ideally we'd run `make maintainer-clean`, which is the expected way of deleting such things (or perhaps even `make bootstrap`). But the tarball only has a `Makefile.in`, so we'd need to run `./configure`, then `make maintainer-clean`, then `./configure` again, which is super ugly. Hopefully deleting the `.elc` files the naïve way is enough?

**Comment by adisbladis** [comment]
@collares I think maybe we should just skip the tarball entirely and fetch it from git instead.

**Comment by thefloweringash** [comment]
> @collares I think maybe we should just skip the tarball entirely and fetch it from git instead.

There are other side-effects caused by skipping the tarball. This [comment by an emacs maintainer](https://old.reddit.com/r/emacs/comments/b2vv47/running_emacs_without_the_fat/ej3rpsf/?context=3) claims, for example, that bootstrapping is only expected to happen to emacs developers. Part of bootstrapping is converting the Japanese input method dictionaries, and this takes some amount of time. They're in the tarball at `lisp/leim/ja-dic/ja-dic.el`, but are absent in the git checkout.

**Comment by adisbladis** [comment]
> There are other side-effects caused by skipping the tarball. This [comment by an emacs maintainer](https://old.reddit.com/r/emacs/comments/b2vv47/running_emacs_without_the_fat/ej3rpsf/?context=1) claims, for example, that bootstrapping is only expected to happen to emacs developers. Part of bootstrapping is converting the Japanese input method dictionaries, and this takes some amount of time. They're in the tarball at `lisp/leim/ja-dic/ja-dic.el`, but are absent in the git checkout.

That's a good point, I would still rather inflate build times a bit and exercise bootstrapping as a part of our regular build though to completely avoid these types of issues.

**Comment by thefloweringash** [comment]
> > There are other side-effects caused by skipping the tarball. This [comment by an emacs maintainer](https://old.reddit.com/r/emacs/comments/b2vv47/running_emacs_without_the_fat/ej3rpsf/?context=1) claims, for example, that bootstrapping is only expected to happen to emacs developers. Part of bootstrapping is converting the Japanese input method dictionaries, and this takes some amount of time. They're in the tarball at `lisp/leim/ja-dic/ja-dic.el`, but are absent in the git checkout.
> 
> That's a good point, I would still rather inflate build times a bit and exercise bootstrapping as a part of our regular build though to completely avoid these types of issues.

I'm hesitant about building things in a way that upstream doesn't expect, unless necessary to adapt for nixpkgs. For a related example, using the repo instead of the tar distribution for Ruby caused a subtle problem where the commit hash was not propagated, which broke the bootsnap gem's cache invalidation (fixed in #161801, by using the tarball). The real question here, I think, is how upstream expects native compilation to be used if their tarball doesn't support it.
